### ISSUE TITLE
Adding support for mongo_engine.MongoEngineStorage

### ISSUE THREAD

**Issue by JohnAD** [ashcrow/flask-track-usage#34]
I'm volunteering to write, test, & document this expansion, but wanted to raise it as a issue first to make sure of no complications.

Essentially, it is not too different than the `MongoPiggybackStorage` variant already written. But it would be referenced the manner someone using the _MongoEngine_ wrapper library would expect. Something like:

```
from flask.ext.track_usage import TrackUsage
from flask.ext.track_usage.storage.mongo_engine import MongoEngineStorage

t = TrackUsage(app, MongoEngineStorage())
```

The only reason MongoEngineStorage would be stored in `.storage.mongo_engine` rather than `.storage.mongo` would be to prevent a needless library dependency for those not using MongoEngine.

If the user were to want to change a characteristic of the storage they would do it with a MongoEngine class inheritance of the default. For example to change the name of the collection:

```
from flask.ext.track_usage import TrackUsage
from flask.ext.track_usage.storage.mongo_engine import MongoEngineStorage, UsageTracker

class MyTrackUsage(UsageTracker):
    meta = {'collection': 'websiteDUsage'}

t = TrackUsage(app, MongoEngineStorage(doc=MyTrackUsage))
```

I also have a second expansion idea, but I'm not sure it is best placed into this library. I'll post a second issue to discuss that.

**Comment by JohnAD** [comment]
Just for reference:
http://mongoengine.org/
https://github.com/MongoEngine/mongoengine
https://github.com/MongoEngine/flask-mongoengine

**Comment by JohnAD** [comment]
For anyone who is a MongoEngine expert, you might have notice what is normally a problem with the above example: one cannot normally change the collection of an inherited document. So, to make the above work, I would need to make some meta-class manipulation work... The final version might well end up slightly different than the example given anyway. :)

**Comment by JohnAD** [comment]
Forked the repo. Made the changes. Will be adding testing and documentation changes next. Should send a Pull request in the next day or so.

**Comment by ashcrow** [comment]
I'm not an expert in MongoEngine but I'm totally ok with this and will accept it 😄

I'd like it in the same ``mongo.py`` file using the same import strategy (import on class instance creation) to keep mongo backends together.

**Comment by JohnAD** [comment]
I've moved it to only import on instance creation. I've also written the nose test.

The MongoEngine document also saves a psuedo "apache combined log" string field as a reference. To do that, I had to add a few more fields to the `data` dictionary created by `TrackUsage.after_request(self, response)`. It shouldn't cause any problems to the other Storage classes.

I'll have something for you to pull to the `2.0.0` branch tomorrow.

BTW, another possible Storage class could, literally, be ApacheLogStorage, where a text file is appended with traditional apache logs.
https://httpd.apache.org/docs/1.3/logs.html

**Comment by ashcrow** [comment]
> I've moved it to only import on instance creation. I've also written the nose test.

👍 

> The MongoEngine document also saves a psuedo "apache combined log" string field as a reference.  To do that, I had to add a few more fields to the data dictionary created by TrackUsage.after_request(self, response). It shouldn't cause any problems to the other Storage classes.

There are some options we could persue for this. We could add a "extra data" field to store unparsed, extra data that may be created ... or we can ignore the result during persistence. Ignoring it would be cleaner, but storing it could be helpful for some people. I don't really have a strong opinion on that.

> I'll have something for you to pull to the 2.0.0 branch tomorrow.

Awesome!

> BTW, another possible Storage class could, literally, be ApacheLogStorage, where a text file is appended with traditional apache logs. https://httpd.apache.org/docs/1.3/logs.html

That's an interesting idea. Providing the metrics back via the API would be extremely hard so it would have to be a write only storage (similar to the ``PrintStorage``). The other issue we may find with this would be multiple processes attempting to write to a single file at the same time (which is solvable for sure). If there is a want/need for it I'm not opposed to including it ... though we may want to somehow flag ``Storage`` classes which are write only so it's obvious that gaining the simple metrics back is not supported. One possible design would be to have ``Storage`` and ``Output`` classes. ``Storage`` would store and be queriable while ``Output`` would have most of the same methods as ``Storage`` **without** the readi/query methods. ``PrintStorage`` would become ``PrintOutput`` and we'd add ``ApacheFileOutput`` and others as requested. Since we would support multiple instances I could see someone wanting to use a ``Storage`` and one or more ``Outputs``.

**tl;dr**
I'm fine with us adding an output style storage but there are some design decisions needed. If you think it's worth it please open a new issue around that and we'll work together to design and implement it.

**Comment by ashcrow** [comment]
Work being done in https://github.com/ashcrow/flask-track-usage/pull/36
